src/Libraries/Microsoft.Extensions.AI.Abstractions/AdditionalPropertiesDictionary{TValue}.cs

```diff
@@ -0,0 +1,249 @@
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Linq;
+using Microsoft.Shared.Diagnostics;
+
+#pragma warning disable S1144 // Unused private types or members should be removed
+#pragma warning disable S2365 // Properties should not make collection or array copies
+#pragma warning disable S3604 // Member initializer values should not be redundant
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>Provides a dictionary used as the AdditionalProperties dictionary on Microsoft.Extensions.AI objects.</summary>
+/// <typeparam name="TValue">The type of the values in the dictionary.</typeparam>
+[DebuggerDisplay("Count = {Count}")]
+[DebuggerTypeProxy(typeof(AdditionalPropertiesDictionary<>.DebugView))]
+public class AdditionalPropertiesDictionary<TValue> : IDictionary<string, TValue>, IReadOnlyDictionary<string, TValue>
+{
+    /// <summary>The underlying dictionary.</summary>
+    private readonly Dictionary<string, TValue> _dictionary;
+
+    /// <summary>Initializes a new instance of the <see cref="AdditionalPropertiesDictionary{TValue}"/> class.</summary>
+    public AdditionalPropertiesDictionary()
+    {
+        _dictionary = new(StringComparer.OrdinalIgnoreCase);
+    }
+
+    /// <summary>Initializes a new instance of the <see cref="AdditionalPropertiesDictionary{TValue}"/> class.</summary>
+    public AdditionalPropertiesDictionary(IDictionary<string, TValue> dictionary)
+    {
+        _dictionary = new(dictionary, StringComparer.OrdinalIgnoreCase);
+    }
+
+    /// <summary>Initializes a new instance of the <see cref="AdditionalPropertiesDictionary{TValue}"/> class.</summary>
+    public AdditionalPropertiesDictionary(IEnumerable<KeyValuePair<string, TValue>> collection)
+    {
+#if NET
+        _dictionary = new(collection, StringComparer.OrdinalIgnoreCase);
+#else
+        _dictionary = new Dictionary<string, TValue>(StringComparer.OrdinalIgnoreCase);
+        foreach (var item in collection)
+        {
+            _dictionary.Add(item.Key, item.Value);
+        }
+#endif
+    }
+
+    /// <summary>Creates a shallow clone of the properties dictionary.</summary>
+    /// <returns>
+    /// A shallow clone of the properties dictionary. The instance will not be the same as the current instance,
+    /// but it will contain all of the same key-value pairs.
+    /// </returns>
+    public AdditionalPropertiesDictionary<TValue> Clone() => new(_dictionary);
+
+    /// <inheritdoc />
+    public TValue this[string key]
+    {
+        get => _dictionary[key];
+        set => _dictionary[key] = value;
+    }
+
+    /// <inheritdoc />
+    public ICollection<string> Keys => _dictionary.Keys;
+
+    /// <inheritdoc />
+    public ICollection<TValue> Values => _dictionary.Values;
+
+    /// <inheritdoc />
+    public int Count => _dictionary.Count;
+
+    /// <inheritdoc />
+    public bool IsReadOnly => false;
```

Comment:
Dogs you purposefully change this to be public rather than explicitly implementing the interface?

Reply:
Yes, after unsealing, it was to deal with:

```
error CA1033: Make 'AdditionalPropertiesDictionary' sealed (a breaking change if this class has previously shipped), implement the method non-explicitly, or implement a new method that exposes the functionality of 'System.Collections.Generic.ICollection<System.Collections.Generic.KeyValuePair<System.String,TValue>>.get_IsReadOnly' and is visible to derived classes (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1033)
```

Same applies to `Contains` and `CopyTo`.

Do you feel we should solve this in another way?

Reply:
> For example I could just suppress S4039 and CA1033 in this file.

I've done that for now. Given that the three members concerned (`IsReadOnly`, `Contains(keyvaluepair)` and `CopyTo(collection of keyvaluepair)`) are all extremely unlikely to be useful on this type, making them hard to access is unlikely to be a problem.

Reply:
I'd just suppress the diagnostic here.